pkg/networkutils/network.go

```diff
@@ -67,28 +74,36 @@ func isDuplicateRuleAdd(err error) bool {
 	return strings.Contains(err.Error(), "File exists")
 }
 
-// SetupNodeNetwork performs node level network configuration
+// SetupHostNetwork performs node level network configuration
 // TODO : implement ip rule not to 10.0.0.0/16(vpc'subnet) table main priority  1024
 func (os *linuxNetwork) SetupHostNetwork(vpcCIDR *net.IPNet, primaryAddr *net.IP) error {
 
+	externalSNAT, err := useExternalSNAT()
+	if err != nil {
+		return err
+	}
+
 	hostRule := os.netLink.NewRule()
 	hostRule.Dst = vpcCIDR
 	hostRule.Table = mainRoutingTable
 	hostRule.Priority = hostRulePriority
 	hostRule.Invert = true
 
-	// if this is a restart, cleanup previous rule first
-	err := os.netLink.RuleDel(hostRule)
+	// If this is a restart, cleanup previous rule first
+	err = os.netLink.RuleDel(hostRule)
 	if err != nil && !containsNoSuchRule(err) {
 		log.Errorf("Failed to cleanup old host IP rule: %v", err)
 		return errors.Wrapf(err, "host network setup: failed to delete old host rule")
 	}
 
-	err = os.netLink.RuleAdd(hostRule)
-
-	if err != nil {
-		log.Errorf("Failed to add host IP rule: %v", err)
-		return errors.Wrapf(err, "host network setup: failed to add host rule")
+	// Only include the rule if SNAT is not being handled by an external NAT gateway and needs to be
+	// handled on-node.
+	if !externalSNAT {
```

Comment:
@liwenwu-amazon :  So strictly this isn't required if you are using external SNAT, but IIRC this is required if you are trying to route from another VPC or across a VPN gateway.  Without it you have the situation:
-  Route to pod from outside VPC (i.e. other VPC or VPN) may come in on eth1 depending on which secondary ENI interface.
-  Route back again is forced over eth0
-  This results in packet drop due to reverse path filtering.

Details are also in #53.

So I feel that it's all related, but perhaps the environment variable is incorrectly named, so rather than focussing on the SNAT aspect, I should maybe focus on the "Allow traffic between VPCs or VPNs" where the additional requirement to turn this feature on is that you also have to use a NAT gateway.

I couldn't come up with a nice succinct name for that though.  Perhaps `AWS_VPC_K8S_CNI_ALLOW_INTER_VPC`?

WDYT?

Reply:
@liwenwu-amazon I've just been retesting on a new EKS cluster and I have a question that is related to all of this.

How do determine which VPC subnet is selected for the pods? 

When I previously tested some of my pods were using eth1, but in my retests they seem to be using eth0 - I've no idea how that came about though and with my latest testing I don't seem to be able to hit this.  Do you know if anything has changed in this area since this PR was created?

Reply:
@robbrockbank right now all ENIs on a node using same PVC subnet and security group as node's primary ENI. In another word, all eth0, eth1 etc. are using same PVC subnet and security group.  ipamdD randomly choose a secondary IP address of these ENIs (e.g. eth0, eth1 etc) and assign it to Pod.

Reply:
if a Pod is assigned with a secondary IP address from eth1,  and if a packet with IP-DA=Pod's address is received on eth1 interface, it will get forwarded to Pod use main routing table(thanks to toPodRulePriority 512 ip rule).

Reply:
what about AWS_VPC_K8S_CNI_DISABLE_NAT ?